HA setup for kubernetes role (multi control-plane support).

bomboclat opened the pull request:
Features
- Adds multi contro-plane join support
- Setup a etcd stacked cluster:
  https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/#stacked-etcd-topology
- Setup a haproxy + keepalived pods cluster:
  https://github.com/kubernetes/kubeadm/blob/main/docs/ha-considerations.md#option-2-run-the-services-as-static-pods

GallAndrea committed:
Edit master task for kuberentes role. Add multi control-plane (master) support.

We added:
- `k8s_multi_master` variable: to check if k8s cluster has multiple
  control-plane nodes
- Credentials generation: these credentials are useful only on multi
  control-plane setup
- `kubeadm init` command: now the command changes if the cluster has
  multiple control-plane nodes
- Added a task block to order the file.

GallAndrea committed:
Add HA setup for kubernetes role

This commit adds the HA cluster setup to the kubernetes role, and it's
complementary to the previous commit: 37fd587
Features added:
- Adds multi contro-plane join support
- Setup a etcd stacked cluster:
  https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/#stacked-etcd-topology
- Setup a haproxy + keepalived pods cluster:
  https://github.com/kubernetes/kubeadm/blob/main/docs/ha-considerations.md#option-2-run-the-services-as-static-pods

GallAndrea committed:
Edit k8s_multi_master variable

Now instead of an integer value, the variable can be true or false.

GallAndrea committed:
Fix worker join command for multi-master infrastructure

GallAndrea committed:
Fix setting up control-plane nodes

Now the control-plane bootstrap it's executed after static pods configuration

rascasoft commented on roles/kubernetes/tasks/master.yml:
```diff
@@ -2,7 +2,7 @@
 - name: Check if the infrastructure is multi-master
   set_fact:
-    k8s_multi_master: "{{ ansible_play_batch | map('extract', hostvars, 'k8s_role') | select('equalto', 'master') | length }}"
+    k8s_multi_master: "{{ ansible_play_batch | map('extract', hostvars, 'k8s_role') | select('equalto', 'master') | length > 1 }}"
```
I need to understand what we're doing here, since the k8s_multi_master variable is crucial for the cluster configuration.
My suggestion is to make k8s_multi_master a boolean defaulted to false. If the user wants to do a clustered configuration, then an override will be created, so please remove the usage of ansible_play_patch here, which is pure over complexity.
Less is more, and instead of counting the number of declared masters we want the user to set:
```yaml
k8s_multi_master: true
```
And we're good!

rascasoft commented on roles/kubernetes/tasks/master.yml:
```diff
@@ -53,12 +53,12 @@
         --pod-network-cidr={{ k8s_network_cidr }}
       become: true
-      when: k8s_multi_master|int > 1
+      when: k8s_multi_master
```
Remember that in ansible when you want to check a boolean best practices suggest:
```yaml
when: k8s_multi_master | bool
```

GallAndrea committed:
Fix k8s_multi_master variable

Following the review process we decided to make k8s_multi_master a
boolean variable, defaulted to false. If the user wants to do a clustered
configuration, then an override will be created, so we remove the
usage of ansible_play_patch, which is pure over complexity.

rascasoft committed:
Kubernetes: fix undefined run_non_infra_pods var

When not defined run_non_infra_pods was causing issues during runs,
this commit fixes the problem by checking first its definition.

rascasoft committed:
Kubernetes: fix HA static pods

This commit fixes the way static pods are deployed by changing the
paths of the configuration files (which are mountPath volumes) and
by fixing the haproxy configuration to correctly route the https
traffic.

rascasoft committed:
Kubernetes: execute users and cephs on main master

This commit adds this condition:

    - inventory_hostname == k8s_master_node

for the tasks involving users and ceph, since they're supposed to
be executed just on the main master node (k8s_master_node).

rascasoft committed:
Kubernetes: separate master vs multi mastr tasks

This commit puts the required files creation, needed on the multi
master setup, to be executed just on the main master node.
It also fixes indentation.

rascasoft committed:
Kubernetes: enable 8443 port in firewalld

By default we're now using 8443 port for haproxy load balancer in
multi master setup, this commit enables the port in
k8s_master_ports defaults variable.

rascasoft committed:
Kubernetes: modify inventory lab for multi master

This commit adds the multi master defaults to the lab configuration.

rascasoft committed:
Kubernetes: put multi master stuff in master.yml

This commit removes multi-master.yml since it ended with just two
actions that is useful to have in master.yml.

rascasoft committed:
Kubernetes: better multi master file generation

This file optimize the way files related to multi master are
produced.

rascasoft committed:
Kubernetes: uniform kubeadm init operations

This commit extends the descriptions inside the defaults main file,
uniforms the way init operations are made in the various modes
(single or multi master) and simplifies the management of the
passwords and certificate keys needed for the Kubernetes cluster
initialization.
